File: LettercasePercentageRatio/Perl/lettercase_percentage_ratio.pl
# lettercase percentage ratio solution in perl for code eval by steven a dunn

use 5.014;
use strict;

my $file = $ARGV[0];
open(INFILE, "<$file") or die("Cannot open file $file for reading: $!");

while(my $line = <INFILE>) {
    # skip blank lines
    next if($line =~ m/^s$/);
    chomp($line);

    my $upper_count = 0;
    my $lower_count = 0;

    my @chars = split(//, $line);
    map { $_ =~ /[A-Z]/ ? $upper_count++ : $lower_count++; } @chars;

    my $total_count = $upper_count + $lower_count;
    my $upper_percent = $upper_count / $total_count * 100;
    my $lower_percent = $lower_count / $total_count * 100;

    my $result = sprintf("lowercase: %.2f uppercase: %.2f \n", $lower_percent, $upper_percent);
    print $result;
}
close(INFILE);

File: LettercasePercentageRatio/c#/LettercasePercentageRatio.cs
// lettercase percentage ratio soln in c# for code eval by steven a dunn

using System;
using System.IO;

class DataRecovery {
  public static void Main(string[] args) {
    using (StreamReader reader = new StreamReader(args[0])) {
      while (!reader.EndOfStream) {
        string line = reader.ReadLine();
        if (null == line)
          continue;

        float upper_count = 0.0f;
        float lower_count = 0.0f;

        char[] chars = line.Trim().ToCharArray();
        foreach (char c in chars) {
          if (char.IsUpper(c))
            upper_count += 1.0f;
          else if (char.IsLower(c))
            lower_count += 1.0f;
        }

        float total_count = upper_count + lower_count;
        float upper_percent = (upper_count / total_count) * 100.0f;
        float lower_percent = (lower_count / total_count) * 100.0f;

        Console.WriteLine("lowercase: " + lower_percent.ToString("N" + 2) +  " uppercase: " + upper_percent.ToString("N" + 2));
      }
    }
  }
}